### DISCUSSION TITLE
Release model and when to PR

### DISCUSSION THREAD

**Discussion by maurolepore** [2DegreesInvesting/practices#4] [Q&A]
I noticed the RStudio folks push most commits directly to master -- not via a PR. They may work on a branch but merge it locally, check, merge, and push. Pick any hot package, say, dplyr, and notice that the number of commits is much higher than the number of closed + open pull requests. So I asked.

Me:

> Do you have any guideline to share for deciding when a maintainer should just merge their own commits directly into the main branch versus via a PR?

RStudio engineer:

> I think it depends very much on your release model. For most of my R packages, the master/main branch on GH is the development branch, and CRAN is the "release branch". So it does not bother me much if I temporarily break the CI, and I often don't do a PR. Of course bigger features, or code that needs to be reviewed are better in PRs.

--

This is a conversation we should probably have between the Git and GitHub “practices” teams. 

a. Release = CRAN; devel = master (CRAN model)
b. Release = latest tag/GH-release; devel = master branch (GitHub workflow model)
c. Release = master branch; devel = devel branch (Gitflow model)

1. What release model(s) should we adopt? 

2. When do we expect PRs?

**Comment by cjyetman** [comment]
This is an interesting, maybe important discussion to have.

I'm not sure I agree with the implication of Gabor's comment. If you have a CRAN release version equivalent to some tag in GitHub, and the dev version that you might expect people to install with `devtools::install_github("2DegreesInvesting/packagefoo")`, that does not preclude one from using PRs exclusively to merge... i.e. I don't think your release model makes much difference.... unless master is always equivalent to your release version and then maybe you at least have a much higher need for PRs.

**Reply by maurolepore** [reply]
> [CRAN + master] does not preclude one from using PRs exclusively to merge.

That's right, but often a PR feels like an overkill -- e.g. for typos. The takeaway, I think, is that maintainers of even high-stakes packages don't PR or review every commit to the development branch -- which maybe `master`/`main` -- depending on your release model.

**Reply by cjyetman** [reply]
yeah, I get that... maybe I fear the precedent that it sets, and then the mental gymnastics one has to do to determine whether or not they actually need to do a PR or not, unless the rule is "always" or some other very small, very clear set of rules
